### DISCUSSION TITLE
How to generate blank objects

### DISCUSSION THREAD

**Discussion by praveenfun2** [instancio/instancio#880] [Q&A] [answered]
Is there a way by which we can disable auto populating fields with random values. 
I have a project where we already have a large number of test cases where we rely only on manually set fields in objects. We don't want any randomly generated values.

**Comment by armandino** [comment]
If I understand correctly, you have an object pre-populated with hard-coded values, e.g.

```java
Person p = new Person();
p.setName("test-name");
// etc ...
```

You don't want any other fields populated with random values. What are you trying to do with this populated object using the library?  Can you provide a specific example?

**Reply by praveenfun2** [reply]
I want to use this library to create models and extend other models with it, so that the values could be reused.
Also, our existing test case uses https://github.com/six2six/fixture-factory.
Because of which we are not creating objects by setters but rather with field names like 
```
Fixture.of(Client.class).addTemplate("valid", new Rule(){{
	add("id", "abcd");
}});
```
We want to replace six2six/fixture-factory with an alternative.

**Reply by armandino** [reply]
**Update:**  see the follow up comment below.

-------
That's an interesting use case. It's not supported out of the box. The only approach I can think of is defining a generator that returns a given object with `AfterGenerate.APPLY_SELECTORS` hint. The hint means that the library will only populate fields if the user provides a selector:

```java

class PopulatedInstanceGenerator<T> implements Generator<T> {
    private final T value;

    PopulatedInstanceGenerator(T value) {
        this.value = value;
    }

    @Override
    public T generate(final Random random) {
        return value;
    }

    @Override
    public Hints hints() {
        return Hints.afterGenerate(AfterGenerate.APPLY_SELECTORS);
    }
}

@SuppressWarnings("unchecked")
public static <T> Model<T> objectModel(T object) {
    return (Model<T>) Instancio.of(object.getClass())
            .generate(root(), new PopulatedInstanceGenerator<>(object))
            .toModel();
}
```

Then you can use it like this:

```java
@Test
void example() {
    Person originalPerson = Person();
    originalPerson.setName("test-name");

    Person person = Instancio.of(objectModel(originalPerson))
            .set(field("age"), 25) // customise fields via selectors
            .create();

    assertThat(person.getName()).isEqualTo("test-name");
    assertThat(person.getAge()).isEqualTo(25);
}
```

**Comment by armandino** [comment, marked as answer]
Next release `4.7.0` will have support for creating blank objects:

https://www.instancio.org/user-guide/#creating-blank-objects
